modin/data_management/query_compiler/base_query_compiler.py

```diff
@@ -0,0 +1,1320 @@
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import pandas
+
+from modin.error_message import ErrorMessage
+from modin.engines.base.block_partitions import BaseBlockPartitions
+
+class BaseQueryCompiler(object):
+    """Abstract Class that handles the queries to Modin dataframes.
+
+    Note: See the Abstract Methods and Fields section immediately below this
+        for a list of requirements for subclassing this object.
+    """
+
+    # Abstract Methods and Fields: Must implement in children classes
+    # In some cases, there you may be able to use the same implementation for
+    # some of these abstract methods, but for the sake of generality they are
+    # treated differently.
+    def __init__(self, block_partitions_object: BaseBlockPartitions, index, columns, dtypes=None,):
+        raise NotImplementedError("Must be implemented in children classes")
+
+    # Dtypes and Indexing Abstract Methods
+    _dtype_cache = None
+
+    def _get_dtype(self):
+        raise NotImplementedError("Must be implemented in children classes")
+
+    def _set_dtype(self, dtypes):
+        self._dtype_cache = dtypes
+
+    dtypes = property(_get_dtype, _set_dtype)
+
+    def compute_index(self, axis, data_object, compute_diff=True):
+        """Computes the index after a number of rows have been removed.
+
+        Note: In order for this to be used properly, the indexes must not be
+            changed before you compute this.
+
+        Args:
+            axis: The axis to extract the index from.
+            data_object: The new data object to extract the index from.
+            compute_diff: True to use `self` to compute the index from self
+                rather than data_object. This is used when the dimension of the
+                index may have changed, but the deleted rows/columns are
+                unknown.
+
+        Returns:
+            A new Index object.
+        """
+        raise NotImplementedError("Must be implemented in children classes")
+
+    # These objects are currently not distributed.
+    _index_cache = None
```

Comment:
As a rule of thumb, we should only really have things in the abstract class that get called from `modin.pandas.DataFrame`. Anything with a leading underscore is just an implementation detail and not needed in the abstract class.